experiments/exp3/ccrypt-1.11/doc/ccrypt.1
.\" Copyright (C) 2000-2018 Peter Selinger.
.\" This file is part of ccrypt. It is free software and it is covered
.\" by the GNU general public license. See the file COPYING for details.
.TH CCRYPT 1 "July 2018" "Version 1.11" "Encryption"
.SH NAME
ccrypt \- encrypt and decrypt files and streams 
.SH SYNOPSIS

.nf
.B ccrypt [\fImode\fP] [\fIoptions\fP] [\fIfile...\fP]
.B ccencrypt [\fIoptions\fP] [\fIfile...\fP]
.B ccdecrypt [\fIoptions\fP] [\fIfile...\fP]
.B ccat [\fIoptions\fP] \fIfile...\fP
.fi
.SH DESCRIPTION

\fBccrypt\fP is a tool for encrypting and decrypting files and
streams. It is based on the Rijndael block cipher, a version
of which is also used in the Advanced Encryption Standard (AES, see
http://www.nist.gov/aes). This cipher is believed to provide very
strong cryptographic security.

The algorithm provided by \fBccrypt\fP is not symmetric, i.e., one
must specify whether to encrypt or decrypt. The most common way to
invoke \fBccrypt\fP is via the commands \fBccencrypt\fP and
\fBccdecrypt\fP.

Encryption and decryption depends on a keyword (or key phrase)
supplied by the user. By default, the user is prompted to enter a
keyword from the terminal. Keywords can consist of any number of
characters, and all characters are significant (although \fBccrypt\fP
internally hashes the key to 256 bits). Longer keywords provide better
security than short ones, since they are less likely to be discovered
by exhaustive search.
.SH MODES

\fBccrypt\fP can operate in five different modes.  If more than one
mode is specified, the last one specified takes precedence. The
aliases \fBccencrypt\fP, \fBccdecrypt\fP, and \fBccat\fP
are provided as a convenience; they are equivalent to \fBccrypt -e\fP,
\fBccrypt -d\fP, and \fBccrypt -c\fP, respectively.
.TP 15
.B -e, --encrypt
Encrypt. This is the default mode. If filename arguments are given,
encrypt the files and append the suffix \fB.cpt\fP to their
names. Otherwise, run as a filter. 
.TP
.B -d, --decrypt
Decrypt. If filename arguments are given, decrypt the files and strip
the suffix \fB.cpt\fP from the filenames, if present. Otherwise, run
as a filter.
.TP
.B -c, --cat
Decrypt one or more files to standard output. If no filename arguments
are given, decrypt as a filter. A filename of "-" is interpreted as
reading from standard input. Implies \fB-l\fP.
.TP
.B -x, --keychange
Change the key of encrypted data. In this mode, \fBccrypt\fP prompts
for two passwords: the old one and the new one.  If filename arguments
are given, modify the files. Otherwise, run as a filter.
.TP
.B -u, --unixcrypt
Simulate the \fBcrypt\fP(1) command-line encryption tool that used to
exist on very old Unix systems from about 1979 until the 1990s. Note:
the cipher used by the old \fBcrypt\fP(1) tool is not secure and has
been broken. This option is provided only for decrypting existing
files.  If filename arguments are given, decrypt the files to
stdout. Otherwise, run as a filter. A filename of "-" is interpreted
as reading from standard input. Since there is no easy way to detect
whether a given key matches or not, for safety, this mode does not
overwrite files.
.PD
.SH OPTIONS

The following options are supported in addition to the modes described
above:
.TP 15
.B -b, --brave
When reading an encryption key from the terminal, ask the user only
once for the key. By default, \fBccrypt\fP will ask the user to enter
such keys twice, as a safeguard against accidentally destroying data
due to a mistyped key. Using the \fB--brave\fP option disables this
safety check. Never use it, unless you know what you are doing. See
also \fB--timid\fP.
.TP
.B -E \fIvar\fP, --envvar \fIvar\fP
Read the keyword from the environment variable \fIvar\fP.
Note that this might be unsafe on certain systems, namely where
users can use the \fBps\fP command to see the environment of a process
owner by another user. On most modern systems, however, such behavior
of \fBps\fP is disabled and the \fB-E\fP option should be safe there.
Also, as an added security measure, \fBccrypt\fP erases the keyword
from its environment immediately after reading it.
.TP
.B -f, --force
Overwrite existing files or operate on write-protected files without
asking any questions. Also, override \fBccrypt\fP's reluctance to
write or read encrypted data to or from a terminal. 
.TP
.B -F \fIvar\fP, --envvar2 \fIvar\fP
Same as \fB-E\fP, except for second keyword (in keychange mode).
.TP
.B -h, --help
Help. Print usage information and exit.
.TP
.B -H \fIkey\fP, --key2 \fIkey\fP
Same as \fB-K\fP, except for second keyword (in keychange mode).
.TP
.B -k \fIfile\fP, --keyfile \fIfile\fP
Read the keyword as the first line from the named file. In keychange
mode, two keywords are read as the first two lines of the file. The
filename "-" may be given for reading keywords from standard input.
Using the \fB-k -\fP option and sending the keyword on stdin is
probably the safest way to pass a keyword to \fBccrypt\fP from another
program or script.
.TP
.B -K \fIkey\fP, --key \fIkey\fP
Specify the keyword on the command line. This is unsafe, because any
other user can see the command line by running the \fBps\fP command.
Only use this option for testing purposes, and never with a real
keyword.
.TP
.B -y \fIfile\fP, --keyref \fIfile\fP
In encryption or keychange mode, check the encryption key against the
named file, which must have been previously encrypted with the same
key. Exit with an error message if the key does not match. This option
is useful as an alternative to \fB--timid\fP, to guard against
mistyped keys in situations where several files are encrypted with the
same key. This option implies \fB--brave\fP, unless the \fB--timid\fP 
option is explicitly given after the \fB--keyref\fP option.
.TP
.B -l, --symlinks
Force encryption/decryption of symbolic links. By default, symbolic
links are ignored except in cat or unixcrypt mode. Note
that with the \fB-l\fP option, encryption/decryption of a symbolic
link causes the suffix \fB.cpt\fP to be added/removed from the name
of the link, not the name of the file pointed to.
.TP
.B -L, --license
Print license info and exit.
.TP
.B -m, --mismatch
Normally, \fBccrypt\fP refuses to decrypt data with a key that does
not seem to match. The \fB-m\fP option overrides this restriction. This
can sometimes be useful in recovering data from a corrupted file (see
RECOVERING DATA FROM CORRUPTED FILES). To avoid irretrievable loss
of data when decrypting with a wrong key, this option cannot be used
with modes that overwrite the input file.
.TP
.B -P \fIprompt\fP, --prompt \fIprompt\fP
Use \fIprompt\fP instead of the default prompt "Enter
encryption/decryption key: ".  This may be useful in some shell scripts.
.TP
.B -q, --quiet
Suppress most warnings.
.TP
.B -Q \fIprompt\fP, --prompt2 \fIprompt\fP
Same as \fB-P\fP, except for second keyword (in keychange mode).
.TP
.B -r, --recursive
Traverse subdirectories recursively.
.TP
.B -R, --rec-symlinks
Traverse subdirectories recursively, and also follow symbolic links to
subdirectories.
.TP
.B -s, --strictsuffix
Refuse to encrypt files that already have the \fB.cpt\fP suffix (or
that selected with \fB-S\fP). This can be useful when adding some
files to a directory of already encrypted files. This option has no
effect in decryption or keychange mode.
.TP
.B -S \fI.suf\fP, --suffix \fI.suf\fP
Use the suffix \fI.suf\fP instead of the default suffix \fB.cpt\fP.
.TP
.B -t, --timid
When reading an encryption key from the terminal, ask the user to
enter the key twice. If the two entered keys are not identical, abort.
This is a safeguard against accidentally destroying data by encrypting
it with a mistyped key. Note: this behavior is now the default, and
can be overridden with the \fB--brave\fP option.
.TP
.B -T, --tmpfiles
This option causes \fBccrypt\fP to use temporary files during
encryption/decryption, rather than overwriting the file contents
destructively. This 
method leaves the original file contents lying around in unused
sectors of the file system, and thus is less secure than the default
behavior. However, in situations where this loss of security is not
important, the \fB--tmpfiles\fP option can provide a measure of
protection against data being corrupted due to a system crash in the
middle of overwriting a file.
.TP
.B -v, --verbose
Print progress information to stderr.
.TP
.B -V, --version
Print version info and exit.
.TP
.B --
End of options. Any remaining arguments are interpreted as filenames.
This also turns off filter mode, even if zero filenames follow. This
might be useful in the context of shell pattern expansion;
\fBccrypt -- *\fP will behave correctly even if no files match the
pattern \fB*\fP.
.PD
.SH NOTES ON USAGE

The user interface of \fBccrypt\fP intentionally resembles that of GNU
\fBgzip\fP, although it is not identical. When invoked with filename
arguments, \fBccrypt\fP normally modifies the files in place,
overwriting their old content. Unlike \fBgzip\fP, the output is not
first written to a temporary file; instead, the data is literally
overwritten. For encryption, this is usually the desired behavior,
since one does not want copies of the unencrypted data to remain in
hidden places in the file system. The disadvantage is that if
\fBccrypt\fP is interrupted in the middle of writing to a file, the
file will end up in a corrupted, partially encrypted state.  However,
in such cases it is possible to recover most of the data; see
RECOVERING DATA FROM CORRUPTED FILES below. If you want to force
\fBccrypt\fP to use temporary files, use the \fB--tmpfiles\fP option. 

When \fBccrypt\fP receives an interrupt signal (Ctrl-C) while updating
a file in place, it does not exit immediately, but rather delays the
exit until after it finishes writing to the current file. This is to
prevent files from being partially overwritten and thus corrupted. If
you want to force \fBccrypt\fP to exit immediately, just press Ctrl-C
twice quickly.

The encryption algorithm used by \fBccrypt\fP uses a random seed that
is different each time. As a result, encrypting the same file twice
will never yield the same result. The advantage of this method is that
similarities in plaintext do not lead to similarities in ciphertext;
there is no way of telling whether the content of two encrypted files
is similar or not.

Because of the use of a random seed, decrypting and re-encrypting a
file with the same key will not lead to an identical file. It is
primarily for this reason that \fBccrypt\fP refuses to decrypt files
with a non-matching key; if this were allowed, there would be no way
afterwards to restore the original file, and the data would be
irretrievably lost.

When overwriting files, special care is taken with hard links and
symbolic links. Each physical file (i.e., each inode) is processed at
most once, no matter how many paths to it are encountered on the
command line or in subdirectories traversed recursively. For each file
that has multiple hard links, a warning is printed, to alert the user
that not all paths to the file might have been properly renamed.
Symbolic links are ignored except in cat mode, or unless the \fB-l\fP
or \fB-R\fP option is given.

Unlike \fBgzip\fP, \fBccrypt\fP does not complain about files that
have improper suffixes. It is legal to doubly encrypt a file. It is
also legal to decrypt a file that does not have the \fB.cpt\fP
suffix, provided the file contains valid data for the given decryption
key. Use the \fB--strictsuffix\fP option if you want to prevent
\fBccrypt\fP from encrypting files that already have a \fB.cpt\fP
suffix. 

Regarding encryption and compression: encrypted data is statistically
indistinguishable from random data, and thus it cannot be
compressed. But of course it is possible to compress the data first,
then encrypt it. Suggested file suffixes are \fB.gz.cpt\fP 
or \fB.gzc\fP.
.SH RECOVERING DATA FROM CORRUPTED FILES

Encrypted data might be corrupted for a number of reasons. For
instance, a file might have been partially encrypted or decrypted if
\fBccrypt\fP was interrupted while processing the file. Or data might
be corrupted by a software or hardware error, or during transmission
over a network. The encryption algorithm used by \fBccrypt\fP is
designed to allow recovery from errors. In general, only a few bytes
of data will be lost near where the error occurred.

Data encrypted by \fBccrypt\fP can be thought of as a sequence of
32-byte blocks. To decrypt a particular block, \fBccrypt\fP only needs
to know the decryption key, the data of the block itself, and the data
of the block immediately preceding it. \fBccrypt\fP cannot tell
whether a block is corrupted or not, except the very first block,
which is special. Thus, if the encrypted data has been altered in the
middle or near the end of a file, \fBccrypt\fP can be run to decrypt
it as usual, and most of the data will be decrypted correctly, except
near where the corruption occurred.

The very first block of encrypted data is special, because it does not
actually correspond to any plaintext data; this block holds the random
seed generated at encryption time. \fBccrypt\fP also uses the very
first block to decide whether the given keyword matches the data or
not. If the first block has been corrupted, \fBccrypt\fP will likely
decide that the keyword does not match; in such cases, the \fB-m\fP
option can be used to force \fBccrypt\fP to decrypt the data anyway.

If a file contains some encrypted and some unencrypted data, or data
encrypted with two different keys, one should decrypt the entire file
with each applicable key, and then piece together the meaningful parts
manually.

Finally, decryption will only produce meaningful results if the data
is aligned correctly along block boundaries. If the block boundary
information has been lost, one has to try all 32 possibilities.
.SH DESCRIPTION OF THE CIPHER

\fIBlock ciphers\fP operate on data segments of a fixed length.  For
instance, the Rijndael block cipher used in \fBccrypt\fP has a block
length of 32 bytes or 256 bits. Thus, this cipher encrypts 32 bytes
at a time.

\fIStream ciphers\fP operate on data streams of any length. There are
several standard modes for operating a block cipher as a stream
cipher. One such standard is \fICipher Feedback (CFB)\fP, defined in
NIST Special Publication 800-38A and ANSI X3.106-1983. \fBccrypt\fP
implements a stream cipher by operating the Rijndael block cipher in
CFB mode.

Let \fIP[i]\fP and \fIC[i]\fP be the \fIi\fPth block of
the plaintext and ciphertext, respectively. CFB mode specifies that 
.IP
.nf
.I C[i] = P[i] ^ E(k,C[i-1])
.fi
.LP
Here \fI^\fP denotes the bitwise exclusive or function, and
\fIE(k,x)\fP denotes the encryption of the block \fIx\fP under the key
\fIk\fP using the block cipher. Thus, each block of the ciphertext is
calculated from the corresponding block of plaintext and the previous
block of ciphertext. Note that in fact, each byte of \fIP[i]\fP can
be calculated from the corresponding byte of \fIC[i]\fP, so that the
stream cipher can be applied to one byte at a time. In particular, the
stream length need not be a multiple of the block size.

Assuming that blocks are numbered starting from \fI0\fP, a special
"initial" ciphertext block \fIC[-1]\fP is needed to provide the base
case for the above formula. This value \fIC[-1]\fP is called the
\fIinitialization vector\fP or \fIseed\fP. The seed is chosen at
encryption time and written as the first block of the encrypted
stream. It is important that the seed is unpredictable; in particular,
the same seed should never by used more than once.  Otherwise, the two
resulting ciphertext blocks \fIC[0]\fP could be related by a simple
xor to obtain information about the corresponding plaintext blocks
\fIP[0]\fP.  If unpredictable seeds are used, CFB is provably as
secure as the underlying block cipher.

In \fBccrypt\fP, the seed is constructed as follows: first, a nonce is
contructed by hashing a combination of the host name, current time,
process id, and an internal counter into a 28-byte value, using a
cryptographic hash function. The nonce is combined with a fixed
four-byte "magic number", and the resulting 32-byte value is encrypted
by one round of the Rijndael block cipher with the given key. This
encrypted block is used as the seed and appended to the beginning of
the ciphertext. The use of the magic number allows \fBccrypt\fP to
detect non-matching keys before decryption.
.SH SECURITY

\fBccrypt\fP is believed to provide very strong cryptographic
security, equivalent to that of the Rijndael cipher with 256-bit block
size and 256-bit key size. Another version of the Rijndael cipher
(with a smaller block size) is used in the U.S. government's Advanced
Encryption Standard (AES, see http://www.nist.gov/aes). Therefore,
this cipher is very well studied and subject to intensive public
scrutiny.  This scrutiny has a positive effect on the cipher's
security. In particular, if an exploitable weakness in this cipher
were ever discovered, this would become widely publicized.

In practical terms, the security of \fBccrypt\fP means that, without
knowledge of the encryption key, it is effectively impossible to
obtain any information about the plaintext from a given
ciphertext. This is true even if a large number of
plaintext-ciphertext pairs are already known for the same
key. Moreover, because \fBccrypt\fP uses a key size of 256 bits, an
exhaustive search of the key space is not feasible, at least as long
as sufficiently long and hard-to-guess keys are actually used in
practice. No cipher is secure if users choose insecure keywords.

On the other hand, \fBccrypt\fP does not attempt to provide 
\fIdata integrity\fP, i.e., it will not attempt to detect whether the
ciphertext was modified after encryption. In particular, encrypted
data can be truncated, leaving the corresponding decrypted data also
truncated, but otherwise consistent. If one needs to ensure data
integrity as well as secrecy, this can be achieved by other
methods. The recommended method is to prepend a cryptographic hash
(for instance, an SHA-1 hash) to the data before encryption.

\fBccrypt\fP does not claim to provide any particular safeguards
against information leaking via the local operating system. While
reasonable precautions are taken, there is no guarantee that
keywords and plaintexts have been physically erased after encryption
in completed; parts of such data might still exist in memory or on
disk. \fBccrypt\fP does not currently use privileged memory pages.

When encrypting files, \fBccrypt\fP by default accesses them in
read-write mode. This normally causes the original file to be
physically overwritten, but on some file systems, this might not be
the case.

Note that the use of the \fB-K\fP option is unsafe in a multiuser
environment, because the command line of a process is visible to other
users running the \fBps\fP command. The use of the \fB-E\fP option is
potentially unsafe for the same reason, although recent versions of
\fBps\fP don't tend to display environment information to other users.
The use of the \fB-T\fP option is unsafe for encryption because the
original plaintext will remain in unused sectors of the file system.
.SH EMACS PACKAGE

There is an emacs package for reading and writing encrypted files.
(Note that this package currently only works with emacs, not with
xemacs.)  This package hooks into the low-level file I/O functions of
emacs, prompting the user for a password where appropriate. It is
implemented in much the same way as support for compressed files. If
you have both the ps-ccrypt and jka-compr packages installed, emacs
can open encrypted files and compressed files; however, it does not
currently work for files that are encrypted and compressed.

To use the package, simply load ps-ccrypt, then edit as usual.
When you open a file with the ".cpt" extension, emacs will prompt you
for a password for the file. It will remember the password for the
buffer, and when you save the file later, it will be automatically
encrypted again (provided you save it with a ".cpt" extension). Except
for the password prompt, the operation of the package should be
transparent to the user. The command M-x ccrypt-set-buffer-password can
be used to change the current password of a buffer.

The simplest way to use this package is to include the lines 
.IP 
.nf
(setq load-path (cons "\fIpath\fP" load-path)) 
(require 'ps-ccrypt "ps-ccrypt.el")
.fi 
.LP 
in your .emacs file, where \fIpath\fP is the directory that holds the
file ps-ccrypt.el.

Limitations of the emacs package: there is no guarantee that
unencrypted information cannot leak to the file system; in fact, the
package sometimes writes unencrypted data to temporary files. However,
auto-saved files are normally treated correctly (i.e., encrypted).
For details, see the comments in the file ps-ccrypt.el.
.SH EXIT STATUS


experiments/exp3/ccrypt-1.11/src/ccrypt.c
/* Copyright (C) 2000-2018 Peter Selinger.
   This file is part of ccrypt. It is free software and it is covered
   by the GNU general public license. See the file COPYING for details. */

/* ccrypt.c: high-level functions for accessing ccryptlib */

/* ccrypt implements a stream cipher based on the block cipher
   Rijndael, the candidate for the AES standard. */

#ifdef HAVE_CONFIG_H
#include <config.h>  /* generated by configure */
#endif

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>

#include "main.h"
#include "ccryptlib.h"
#include "unixcryptlib.h"
#include "platform.h"

#include "gettext.h"
#define _(String) gettext (String)
#define N_(String) gettext_noop (String)

/* heuristically, the fastest inbufsize is 992 - this is slightly, but
   significantly, faster than 1024 or very large buffer sizes. I'm not
   sure why this is - maybe some strange interaction between the
   filesystem blocksize and the page size? */

#define INBUFSIZE 992  
#define MIDBUFSIZE 1024  /* for key change */
#define OUTBUFSIZE 1056

/* ---------------------------------------------------------------------- */
/* ccrypt error messages. These correspond to the error codes returned
   by the ccrypt functions. Note: the error code -3 corresponds to an
   i/o error, with errno set, the error code -1 corresponds to any
   other system error, with errno set, and the error code -2
   corresponds to a ccrypt error, with ccrypt_errno set. */

/* ---------------------------------------------------------------------- */
/* error handling */

const char *ccrypt_error(int st) {
  if (st == -1 || st == -3) {
    return strerror(errno);
  } 
  if (st == -2) {
    switch (ccrypt_errno) {
    case CCRYPT_EFORMAT:
      return _("bad file format");
      break;
    case CCRYPT_EMISMATCH:
      return _("key does not match");
      break;
    case CCRYPT_EBUFFER:
      return _("buffer overflow");
      break;
    default:
      /* do nothing */
      break;
    }
  }
  return _("unknown error");
}

/* ---------------------------------------------------------------------- */

/* ccdecrypt_init_r is a variant of ccdecrypt_init that works around
   bug #1872759. An unlucky combination of a past version of ccrypt,
   cygwin, and windows accidentally read the key file in binary mode,
   resulting in '\r' to be appended to the end of the key string.
   While the current version of xreadline() strips '\r' from the end
   of keys read from a file, we must still support decryption of these
   legacy files for backward compatibility. To this end,
   ccdecrypt_init_r registers two alternate keys: "key" and "key\r".
   The first matching one is used. */

static int ccdecrypt_init_r(ccrypt_stream_t *b, const char *key, int flags) {
  char *key2;
  const char *keylist[2];
  int len = strlen(key);
  int r;

  key2 = (char *)malloc(len+2);
  if (!key2) {
    return -1;
  }
  strcpy(key2, key);
  key2[len] = '\r';
  key2[len+1] = 0;

  keylist[0] = key;
  keylist[1] = key2;
  r = ccdecrypt_multi_init(b, 2, keylist, flags);
  free(key2);
  return r;
}


/* ---------------------------------------------------------------------- */
/* keychange = compose decryption and encryption */

struct keychange_state_s {
  ccrypt_stream_t b1;
  ccrypt_stream_t b2;
  int iv;                /* count-down for IV bytes */
  char buf[MIDBUFSIZE];
};
typedef struct keychange_state_s keychange_state_t;

static void keychange_state_free(keychange_state_t *st) {
  ccdecrypt_end(&st->b1);
  ccencrypt_end(&st->b2);
  free(st);
}

static int keychange_init(ccrypt_stream_t *b, const char *key1, const char *key2) {
  keychange_state_t *st;
  int r;
  int cerr, err;

  st = (keychange_state_t *)malloc(sizeof(keychange_state_t));
  if (st == NULL) {
    return -1;
  }
  b->state = (void *)st;
  st->b1.state = NULL;
  st->b2.state = NULL;

  r = ccdecrypt_init_r(&st->b1, key1, 0);
  if (r) {
    goto error;
  }
  r = ccencrypt_init(&st->b2, key2);
  if (r) {
    goto error;
  }
  st->b2.next_in = &st->buf[0];
  st->b2.avail_in = 0;
  st->iv = 32;  /* count down IV bytes */

  return 0;

 error:
  cerr = ccrypt_errno;
  err = errno;
  keychange_state_free(st);
  b->state = NULL;
  ccrypt_errno = cerr;
  errno = err;
  return r;
}

static int keychange(ccrypt_stream_t *b) {
  keychange_state_t *st = (keychange_state_t *)b->state;
  int r;
  int cerr, err;

  /* note: we do not write anything until we have seen 32 bytes of
     input. This way, we don't write the output IV until the input IV
     has been verified. */

  while (1) { 
    /* clear mid-buffer */
    if (b->avail_out && !st->iv) {
      st->b2.next_out = b->next_out;
      st->b2.avail_out = b->avail_out;
      r = ccencrypt(&st->b2);
      if (r) {
	goto error;
      }
      b->next_out = st->b2.next_out;
      b->avail_out = st->b2.avail_out;
    }
    
    /* if mid-buffer not empty, or no input available, stop */
    if (st->b2.avail_in != 0 || b->avail_in == 0) {
      break;
    }

    /* fill mid-buffer */
    st->b1.next_out = &st->buf[0];
    st->b1.avail_out = MIDBUFSIZE;
    st->b1.next_in = b->next_in;
    st->b1.avail_in = b->avail_in;
    r = ccdecrypt(&st->b1);
    if (r) {
      goto error;
    }
    if (st->iv) {
      st->iv -= b->avail_in - st->b1.avail_in;
      if (st->iv <= 0) {
	st->iv = 0;
      }
    }
    b->next_in = st->b1.next_in;
    b->avail_in = st->b1.avail_in;
    st->b2.next_in = &st->buf[0];
    st->b2.avail_in = st->b1.next_out - st->b2.next_in;
  }
  return 0;

 error:
  cerr = ccrypt_errno;
  err = errno;
  keychange_state_free(st);
  b->state = NULL; /* guard against double free */
  ccrypt_errno = cerr;
  errno = err;
  return r;
}

static int keychange_end(ccrypt_stream_t *b) {
  keychange_state_t *st = (keychange_state_t *)b->state;
  int r;
  int cerr, err;

  if (st == NULL) {
    return 0;
  }
  r = ccdecrypt_end(&st->b1);
  if (r) {
    goto error;
  }
  r = ccencrypt_end(&st->b2);
  if (r) {
    goto error;
  }
  free(b->state);
  b->state = NULL;
  return 0;

 error:
  cerr = ccrypt_errno;
  err = errno;
  keychange_state_free(st);
  b->state = NULL;
  ccrypt_errno = cerr;
  errno = err;
  return r;
}

/* ---------------------------------------------------------------------- */
/* encryption/decryption of streams */

typedef int initfun(ccrypt_stream_t *b, const char *key);
typedef int workfun(ccrypt_stream_t *b);
typedef int endfun(ccrypt_stream_t *b);

/* apply ccrypt_stream to pipe stuff from fin to fout. Assume the
   ccrypt_stream has already been initialized. */
static int streamhandler(ccrypt_stream_t *b, workfun *work, endfun *end, 
			 FILE *fin, FILE *fout) {
  /* maybe should align buffers on page boundary */
  char inbuf[INBUFSIZE], outbuf[OUTBUFSIZE]; 
  int eof = 0;
  int r;
  int cerr, err;
  int n;

  clearerr(fin);

  b->avail_in = 0;

  while (1) {
    /* fill input buffer */
    if (b->avail_in == 0 && !eof) {
      r = fread(inbuf, 1, INBUFSIZE, fin);
      b->next_in = &inbuf[0];
      b->avail_in = r;
      if (r<INBUFSIZE) {
	eof = 1;
      }
      if (ferror(fin)) {
	r = -3;
	goto error;
      }
    }
    /* prepare output buffer */
    b->next_out = &outbuf[0];
    b->avail_out = OUTBUFSIZE;

    /* do some work */
    r = work(b);
    if (r) {
      return r;
    }
    /* process output buffer */
    if (b->avail_out < OUTBUFSIZE) {
      n = OUTBUFSIZE - b->avail_out;
      r = fwrite(outbuf, 1, n, fout);
      if (r < n) {
	r = -3;
	goto error;
      }
    }
    if (eof && b->avail_out != 0) {
      break;
    }
  }
  r = end(b);
  if (r) {
    return r;
  }
  return 0;

 error:
  err = errno;
  cerr = ccrypt_errno;
  end(b);
  errno = err;
  ccrypt_errno = cerr;
  return r;
}  

int ccencrypt_streams(FILE *fin, FILE *fout, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = ccencrypt_init(b, key);
  if (r) {
    return r;
  }

  return streamhandler(b, ccencrypt, ccencrypt_end, fin, fout);
}

int ccdecrypt_streams(FILE *fin, FILE *fout, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;
  int flags = 0;

  if (cmd.mismatch) {
    flags |= CCRYPT_MISMATCH;
  }

  r = ccdecrypt_init_r(b, key, flags);
  if (r) {
    return r;
  }

  return streamhandler(b, ccdecrypt, ccdecrypt_end, fin, fout);
}

int cckeychange_streams(FILE *fin, FILE *fout, const char *key1, const char *key2) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = keychange_init(b, key1, key2);
  if (r) {
    return r;
  }

  return streamhandler(b, keychange, keychange_end, fin, fout);
}

int unixcrypt_streams(FILE *fin, FILE *fout, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = unixcrypt_init(b, key);
  if (r) {
    return r;
  }

  return streamhandler(b, unixcrypt, unixcrypt_end, fin, fout);
}

/* check if the key matches the given file. This is done by decrypting
   the first 32 bytes, ignoring the output. Return 0 on success, -1 on
   error with errno set, -2 on error with ccrypt_errno set. */
int keycheck_stream(FILE *fin, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;
  char inbuf[32];
  char outbuf[32];
  int cerr, err;

  clearerr(fin);

  r = ccdecrypt_init_r(b, key, 0);
  if (r) {
    return r;
  }

  /* fill input buffer */
  r = fread(inbuf, 1, 32, fin);
  if (ferror(fin)) {
    r = -3;
    goto error;
  }
  b->next_in = &inbuf[0];
  b->avail_in = r;
  b->next_out = &outbuf[0];
  b->avail_out = 32;
  
  /* try to decrypt */
  r = ccdecrypt(b);
  if (r) {
    return r;
  }
  r = ccdecrypt_end(b);
  if (r) {
    return r;
  }
  return 0;
 
 error:
  err = errno;
  cerr = ccrypt_errno;
  ccdecrypt_end(b);
  errno = err;
  ccrypt_errno = cerr;
  return r;
}

/* ---------------------------------------------------------------------- */
/* destructive encryption/decryption of files */

/* A large value of FILEINBUFSIZE keeps the cost of "lseek" down. It
   is okay for FILEINBUFSIZE to be much larger than MIDBUFSIZE.
   FILEOUTBUFSIZE must be large enough to hold the encryption of
   FILEINBUFSIZE in one piece, or otherwise there will be a buffer
   overflow error. */

#define FILEINBUFSIZE 10240
#define FILEOUTBUFSIZE (FILEINBUFSIZE+32)

/* apply ccrypt_stream to destructively update (and resize) the given
   fd, which must be opened in read/write mode and seekable.
   Encryption will begin at the current file position (normally 0),
   and extend until the end of the file. Note: this only works if the
   stream encoder b/work/end expands its input by at most
   FILEINBUFSIZE bytes; otherwise there will be a buffer overflow
   error. */

static int filehandler(ccrypt_stream_t *b, workfun *work, endfun *end,
		       int fd) {
  /* rp = reader's position, wp = writer's position, fp = file position */
  int p;      /* rp-wp */
  char inbuf[FILEINBUFSIZE];
  char outbuf[FILEINBUFSIZE+32];
  int inbufsize, outbufsize;
  off_t offs;
  int r;
  int i;
  int eof=0;
  int err, cerr;

  p = 0;
  outbufsize = 0;

  while (1) {
    /* file is at position wp */
    if (p != 0) {
      r = lseek(fd, p, SEEK_CUR);
      if (r == -1) {
	r = -3;
	goto error;
      }
    }
    /* file is at position rp */

    /* read block */
    i = 0;
    while (i<FILEINBUFSIZE && !eof) {
      r = read(fd, inbuf+i, FILEINBUFSIZE-i);
      if (r == -1) {
	r = -3;
	goto error;
      } else if (r==0) {
	eof = 1;
      }
      i += r;
    }
    p += i;
    inbufsize = i;

    /* file is at position rp */
    if (p != 0) {
      r = lseek(fd, -p, SEEK_CUR);
      if (r == -1) {
	r = -3;
	goto error;
      }
    }
    /* file is at position wp */

    /* write previous block */
    if (outbufsize > p && !eof) {
      ccrypt_errno = CCRYPT_EBUFFER; /* buffer overflow; should never happen */
      r = -2;
      goto error;
    }
    if (outbufsize != 0) {
      i = 0;
      while (i<outbufsize) {
	r = write(fd, outbuf+i, outbufsize-i);
	if (r == -1) {
	  r = -3;
	  goto error;
	}
	i += r;
      }
      p -= outbufsize;
      outbufsize = 0;
    }
    
    /* encrypt block */
    b->next_in = inbuf;
    b->avail_in = inbufsize;
    b->next_out = outbuf;
    b->avail_out = FILEINBUFSIZE+32;

    r = work(b);
    if (r) {
      return r;
    }      
    
    if (b->avail_in != 0) {
      ccrypt_errno = CCRYPT_EBUFFER; /* buffer overflow; should never happen */
      r = -2;
      goto error;
    }
    inbufsize = 0;
    outbufsize = FILEINBUFSIZE+32-b->avail_out;

    if (eof && outbufsize == 0) { /* done */
      break;
    }
  }
  /* file is at position wp */

  /* close the stream (we need to do this before truncating, because
     there might be an error!) */
  r = end(b);
  if (r) {
    return r;
  }

  /* truncate the file to where it's been written */
  r = offs = lseek(fd, 0, SEEK_CUR);
  if (r == -1) {
    return -1;
  }
  r = ftruncate(fd, offs);
  if (r == -1) {
    return -1;
  }
  
  return 0;

 error:
  err = errno;
  cerr = ccrypt_errno;
  end(b);
  errno = err;
  ccrypt_errno = cerr;
  return r;
}

int ccencrypt_file(int fd, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = ccencrypt_init(b, key);
  if (r) {
    return r;
  }

  return filehandler(b, ccencrypt, ccencrypt_end, fd);
}

int ccdecrypt_file(int fd, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = ccdecrypt_init_r(b, key, 0);
  if (r) {
    return r;
  }

  return filehandler(b, ccdecrypt, ccdecrypt_end, fd);
}

int cckeychange_file(int fd, const char *key1, const char *key2) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = keychange_init(b, key1, key2);
  if (r) {
    return r;
  }

  return filehandler(b, keychange, keychange_end, fd);
}

int unixcrypt_file(int fd, const char *key) {
  ccrypt_stream_t ccs;
  ccrypt_stream_t *b = &ccs;
  int r;

  r = unixcrypt_init(b, key);
  if (r) {
    return r;
  }

  return filehandler(b, unixcrypt, unixcrypt_end, fd);
}
